media/js/base/mozilla-traffic-cop.js

```diff
@@ -0,0 +1,174 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// create namespace
+if (typeof Mozilla === 'undefined') {
+    var Mozilla = {};
+}
+
+/**
+ * Traffic Cop traffic redirector for A/B/x testing
+ *
+ * Example usage:
+ *
+ * var cop = new Mozilla.TrafficCop({
+ *     id: 'exp_firefox_new_all_link',
+ *     variations: {
+ *         'v=1': 25,
+ *         'v=2': 25,
+ *         'v=3': 25
+ *     }
+ * });
+ *
+ * cop.init();
+ *
+ *
+ * @param Object config: Object literal containing the following:
+ *      String id (required): Unique-ish string for cookie identification.
+ *          Only needs to be unique to other currently running tests.
+ *      Object variations (required): Object holding key/value pairs of
+ *          variations and their respective traffic percentages. Example:
+ *
+ *          variations: {
+ *              'v=1': 20,
+ *              'v=2': 20,
+ *              'v=3': 20
+ *          }
+ */
+Mozilla.TrafficCop = function(config) {
+    'use strict';
+
+    // make sure config is an object
+    config = (typeof config === 'object') ? config : {};
+
+    // store id
+    this.id = config.id;
+
+    // store variations
+    this.variations = config.variations;
+
+    // store total percentage of users targeted
+    this.totalPercentage = 0;
+
+    this.redirectVariation;
+
+    // calculate and store total percentage of variations
+    for (var v in this.variations) {
+        this.totalPercentage += this.variations[v];
+    }
+
+    return this;
+};
+
+/*
+ * Initialize the traffic cop. Validates variations, ensures user is not
+ * currently viewing a variation, and (possibly) redirects to a variation
+ */
+Mozilla.TrafficCop.prototype.init = function() {
+    var redirectUrl;
+
+    // make sure config is valid (id & variations present)
+    if (this.verifyConfig()) {
+        // make sure current page doesn't match a variation
+        // (to avoid infinite redirects)
+        if (!this.isVariation()) {
+            // roll the dice to see if user should be send to a variation
+            redirectUrl = this.generateRedirectUrl();
+
+            // if we get a variation, send the user
+            if (redirectUrl) {
+                Mozilla.Cookies.setItem(this.id, this.redirectVariation);
+                window.location.href = redirectUrl;
+            }
+        }
+    }
+};
+
+/*
+ * Ensures variations were provided and in total capture between 1 and 99%
+ * of users.
+ */
+Mozilla.TrafficCop.prototype.verifyConfig = function() {
+    if (!this.id) {
+        return false;
+    }
+
+    if (!this.variations) {
```

Comment:
Could we check that `this.variations` is an Object and contains at least one key/value pair?